lib/rubygems/package.rb

```diff
@@ -239,12 +239,14 @@ def add_metadata tar # :nodoc:
   ##
   # Builds this package based on the specification set by #spec=
 
-  def build skip_validation = false
+  def build skip_validation = false, allow_warnings = true
     Gem.load_yaml
     require 'rubygems/security'
 
     @spec.mark_version
-    @spec.validate unless skip_validation
+    unless skip_validation || !@spec.validate(true, allow_warnings)
+      raise Gem::InvalidSpecificationException, "Spec failed to validate"
```

Comment:
It's not a no-op -- it changes the return value of `validate` to be `false` if there are warnings